### ISSUE TITLE
Type unsafety in current type inference

### ISSUE THREAD

**Issue by iluuu1994** [vimeo/psalm#1927]
https://psalm.dev/r/2ccc805ef3

```php
<?php

class Base
{
}

class Derived extends Base
{
}

class Derived2 extends Base
{
}

/**
 * @template T
 */
class Box
{
    /** @var T */
    public $value;

    /**
     * @param T $value
     */
    public function __construct($value)
    {
        $this->value = $value;
    }

    /**
     * @param T $value
     */
    public function set($value)
    {
        $this->value = $value;
    }
}

class Foo
{
    /**
     * @var Box<Derived> $box
     */
    public $box;

    /**
     * @return Box<Base>
     */
    public function bar(): Box
    {
        $box = new Box(new Derived());
        $this->baz($box);
        return $box;
    }

    /**
     * @param Box<Derived> $box
     */
    public function baz($box)
    {
        $this->box = $box;
    }
}

$foo = new Foo();
$baseBox = $foo->bar();
$baseBox->set(new Derived2());

var_dump($foo->box); // Type says `Box<Derived>` but value is `Box<Derived2>`
```

`$this->baz($box);` should probably fix the type of `$box` to `Foo<Derived>`.

**Comment by iluuu1994** [comment]
Simpler example:

https://psalm.dev/r/57663a1f67

```php
<?php

class Base {}
class Derived extends Base {}

/**
 * @template T of Base
 */
class Foo {
    /**
     * @param T $t
     */
    public function __construct ($t) {}
}

/**
 * @return Foo<Base>
 */
function returnFooBase() {
    $f = new Foo(new Derived());
    takesFooDerived($f);
    return $f; // Invalid return type
}

/**
 * @param Foo<Derived> $foo
 */
function takesFooDerived($foo): void {}

var_dump(returnFooBase());
```

**Comment by muglug** [comment]
This specific issue is now fixed

**Comment by iluuu1994** [comment]
It's very easily tricked though: https://psalm.dev/r/4192f651e4

Also, the same thing can happen by assigning to any variable or property. I'm not sure how Hack handles theses cases but it seems pretty hard to get right.

**Comment by muglug** [comment]
@iluuu1994 fixed in 9aa0fd0

**Comment by iluuu1994** [comment]
@muglug Thanks!

As mentioned, same thing should happen with properties: https://psalm.dev/r/75d47e9080
I'm guessing Hack also gives an error in this example?
